addons/sale_loyalty/controllers/portal.py

```diff
@@ -0,0 +1,40 @@
+from werkzeug import urls
+
+from odoo import http, Command
+from odoo.http import request
+
+from odoo.addons.sale.controllers.portal import CustomerPortal as portal
+from odoo.addons.payment import utils as payment_utils
+
+
+class CustomerPortal(portal):
+
+    def _get_order_page_values(self, order_sudo, message=False):
+        values = super()._get_order_page_values(order_sudo, message)
+        values['loyalty_data'] = order_sudo.state == 'sale' and order_sudo.loyalty_data
+
+        return values
+
+    @http.route(['/topup'], type='http', auth="public", website=True, sitemap=False)
+    def topup_wallet(self, product_id):
+        order = request.env['sale.order'].sudo().create({
+            'partner_id': request.env.user.partner_id.id,
+            'order_line': [
+                Command.create({
+                    'product_id': int(product_id),
+                })
+            ]
+        })
```

Comment:
A public route creating a new order in sudo, with any product id given ?
This seems like a nice way to bloat a database :eyes:

![image](https://github.com/user-attachments/assets/9f211325-3725-45cb-b3be-a4a5d75a9126)